File: web/src/views/plan.ts
import { ago, escapeHtml } from "../ui/format.js";

/* ==========================================================================
   The plan — how much of a subscription has gone, which is the one figure in this
   panel that no file on this machine knows. The server runs Claude Code's own
   `/usage` for it, so the reading is the same one the terminal would give you.
   ========================================================================== */
export const planScrim = document.getElementById("planScrim");
const planButton = document.getElementById("planButton");
const planChipText = document.getElementById("planChipText");

/* Five seconds and a process per reading, and a figure that moves in points
   over hours: so it is read on the clock rather than on the poll, and not at
   all while nobody is looking at the page. */
const PLAN_POLL_MS = 300_000;
export let plan = null;
let planBusy = false;
let planPolledAt = 0;
let planGone = false;      // the panel is read-only, so there is nothing to ask

export async function fetchPlan(force) {
  if (planBusy || planGone) return;
  if (!force && document.hidden) return;
  if (!force && Date.now() - planPolledAt < PLAN_POLL_MS) return;
  planBusy = true;
  planPolledAt = Date.now();
  paintPlanChip();
  try {
    const response = await fetch(`/api/plan${force ? "?force=1" : ""}`, { cache: "no-store" });
    if (response.status === 403) {
      // Reading it means running a command, which a panel serving the network
      // does not do. The chip goes rather than sitting there refusing.
      planGone = true;
      planButton.hidden = true;
      return;
    }
    if (!response.ok) throw new Error(String(response.status));
    plan = await response.json();
    // A reading already in flight elsewhere: come back for its answer shortly
    // rather than waiting on this request.
    if (plan.reading) planPolledAt = Date.now() - PLAN_POLL_MS + 4000;
  } catch (error) {
    /* leave the last reading on screen */
  } finally {
    planBusy = false;
    paintPlanChip();
    if (planScrim.dataset.open === "true") paintPlanDialog();
  }
}

/* The two figures worth an app bar: how much of the current session's allowance
   has gone, and how much of the week's. Whichever is tighter colours the chip. */
function planHeadline() {
  const limits = plan?.limits || [];
  const session = limits.find((l) => /session/i.test(l.name));
  const week = limits.find((l) => /week/i.test(l.name) && !/\(/.test(l.name))
    || limits.find((l) => /week.*all models/i.test(l.name))
    || limits.find((l) => /week/i.test(l.name));
  return [session, week].filter(Boolean);
}

function paintPlanChip() {
  if (planGone) { planButton.hidden = true; return; }
  planButton.hidden = false;
  planButton.dataset.reading = planBusy || plan?.reading ? "1" : "0";
  const shown = planHeadline();
  if (!shown.length) {
    planChipText.textContent = plan?.ok === false ? "plan?" : "plan";
    planButton.dataset.tight = "0";
    planButton.title = plan?.message || "How much of your plan has gone";
    return;
  }
  // Used, not left: it reads the same way the bars below it do, and the colour
  // says how worried to be without doing the subtraction in your head.
  planChipText.innerHTML = shown
    .map((l) => `<span class="plan-pct" data-band="${planBand(l.percent)}">${l.percent}%</span>`)
    .join(`<span class="plan-chip__sep">·</span>`);
  const worst = Math.max(...shown.map((l) => l.percent));
  planButton.dataset.tight = String(planBand(worst));
  planButton.title = `${shown.map((l) => `${l.name}: ${l.percent}% used`).join(", ")} — press for the rest`;
}

/* One scale for every plan figure: room, tight, nearly gone. */
const planBand = (percent) => (percent >= 90 ? 2 : percent >= 75 ? 1 : 0);

function planBar(limit) {
  const tight = planBand(limit.percent);
  return `<div class="plan-limit">
    <div class="plan-limit__head">
      <span class="md-body-medium">${escapeHtml(limit.name)}</span>
      <span class="plan-limit__pct plan-pct md-title-medium md-mono"
            data-band="${tight}">${limit.percent}% used</span>
    </div>
    <div class="use-bar">
      <div class="use-bar__fill" style="width: ${limit.percent ? Math.max(1, limit.percent) : 0}%" data-tight="${tight}"></div>
    </div>
    <div class="use-legend md-label-small">
      <span>${limit.resets ? `resets ${escapeHtml(limit.resets)}` : "nothing used yet"}</span>
    </div>
  </div>`;
}

function paintPlanDialog() {
  const head = document.getElementById("planHead");
  const body = document.getElementById("planBody");
  const age = document.getElementById("planAge");
  if (!plan) {
    head.textContent = "Reading your usage…";
    head.hidden = false;
    body.innerHTML = "";
    age.textContent = "";
    return;
  }
  const limits = plan.limits || [];
  // `/usage`'s own headline only ever says you are on a subscription, which the
  // dialog is already about. It is worth the line only when something failed.
  head.textContent = limits.length ? (plan.message || "") : (plan.message || plan.headline || "");
  head.hidden = !head.textContent;
  body.innerHTML = `
    ${limits.length ? limits.map(planBar).join("")
      : plan.text ? `<p class="plan-raw md-body-small">${escapeHtml(plan.text)}</p>`
      : `<p class="md-body-medium">Nothing came back.</p>`}
    ${(plan.blocks || []).map((block) => {
      // Figures only. `/usage` pads each block with a sentence or two of
      // caveat, and a line carrying no number is one of those.
      const lines = block.lines.filter((line) => /\d/.test(line));
      return lines.length ? `<section class="plan-block">
        <h3 class="plan-block__title md-title-small">${escapeHtml(block.title)}</h3>
        <ul class="md-body-small">${lines
          .map((line) => `<li>${escapeHtml(line)}</li>`).join("")}</ul>
      </section>` : "";
    }).join("")}`;
  age.textContent = plan.at
    ? planBusy || plan.reading ? "reading…" : `read ${ago(Date.now() / 1000 - plan.at)}`
    : "";
}

export function openPlan(open) {
  planScrim.dataset.open = String(open);
  if (!open) { planButton.focus(); return; }
  paintPlanDialog();
  document.getElementById("closePlan").focus();
  // Opening it is asking for the figure, so a stale one is refreshed there and
  // then rather than at the next tick of the clock.
  if (!plan || Date.now() / 1000 - (plan.at || 0) > PLAN_POLL_MS / 1000) fetchPlan(true);
}

planButton.addEventListener("click", () => openPlan(true));
document.getElementById("closePlan").addEventListener("click", () => openPlan(false));
document.getElementById("planRefresh").addEventListener("click", () => { fetchPlan(true); paintPlanDialog(); });
planScrim.addEventListener("click", (event) => { if (event.target === planScrim) openPlan(false); });

File: watchtower/plan.py
"""How much of the subscription has gone.

The one figure in the panel that no file on this machine knows: it comes from
running Claude Code's own `/usage`, so the reading is the same one the terminal
would give you. Held for a while and read on the clock rather than on the poll,
because it is an errand rather than a file read.
"""

from __future__ import annotations

import re
import shutil
import subprocess
import threading
import time

from watchtower import config
from watchtower.config import HOME
from watchtower.errands import own_errand
from watchtower.git.message import MESSAGE_MODEL
from watchtower.git.write import git_said


# `/usage` is the one thing in this panel that no file on this machine knows.
# What a subscription has left is the account's, not the session's, and it lives
# behind Anthropic's API — so it is asked for the way you would ask yourself, by
# running Claude Code's own command and reading what it prints.
#
# Which is the point of doing it this way. The alternative was for the panel to
# read the OAuth token out of ~/.claude/.credentials.json and call an undocumented
# endpoint itself: a web server on this machine holding your credentials, for a
# reading the official client already gives away for free. This spawns `claude`,
# handles no secret, and asks for nothing the terminal would not have told you.
#
# It costs no tokens — the command fetches and prints, and samples no model, which
# a run against a fresh transcript confirms: not one usage entry — but it does take
# five seconds and a process, so it is asked rarely and its answer is kept.
PLAN_TIMEOUT = 45.0


PLAN_FRESH = 300.0


PLAN_LOCK = threading.Lock()


PLAN_HELD: dict = {}


# "Current session: 34% used · resets Aug 12, 5:49pm (Europe/Amsterdam)", and the
# week's two lines in the same shape. The reset clause is optional: a limit at 0%
# has nothing to reset from yet.
PLAN_LIMIT = re.compile(
    r"^(?P<name>[^:]{1,60}?):\s*(?P<percent>\d{1,3})%\s*used"
    r"(?:\s*[·|-]\s*resets\s*(?P<resets>.+?))?\s*$")


# "Last 24h · 4141 requests · 46 sessions" — the heading of a block of bullets.
PLAN_BLOCK = re.compile(r"^(Last\s.+|What's contributing.*)$")


def parse_plan(text: str) -> dict:
    """Read `/usage`'s report into figures, keeping the text it came from.

    The output is a human's report rather than an interface, so nothing here
    insists on it. Every line that reads as a limit becomes one; anything else is
    kept in order as prose, and a run that parses to nothing still hands back
    what it was given rather than an empty panel.
    """
    headline, limits, blocks = "", [], []
    for raw in text.splitlines():
        line = raw.rstrip()
        if not line.strip():
            continue
        match = PLAN_LIMIT.match(line.strip())
        if match:
            limits.append({
                "name": match.group("name").strip(),
                "percent": min(100, int(match.group("percent"))),
                "resets": (match.group("resets") or "").strip(),
            })
            continue
        if PLAN_BLOCK.match(line.strip()):
            blocks.append({"title": line.strip(), "lines": []})
            continue
        if blocks and raw.startswith((" ", "\t")):
            blocks[-1]["lines"].append(line.strip())
            continue
        if not headline:
            headline = line.strip()
        elif blocks:
            blocks[-1]["lines"].append(line.strip())
    return {"headline": headline, "limits": limits, "blocks": blocks, "text": text.strip()}


def run_plan() -> dict:
    """Run `claude /usage` once and read the answer."""
    claude = shutil.which("claude")
    if not claude:
        return {"ok": False, "message": "Cannot find the claude command on PATH"}
    try:
        # Same shape as the commit-message errand: printing, no tools, so there is
        # no permission prompt to answer and nothing it can touch. Run from home
        # rather than a repository — this is the account's reading, not a folder's.
        process = subprocess.Popen(
            [claude, "--print", "/usage", "--model", MESSAGE_MODEL,
             "--allowed-tools", "", "--output-format", "text"],
            cwd=str(HOME), text=True,
            stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
        )
    except (OSError, subprocess.SubprocessError) as error:
        return {"ok": False, "message": f"Could not run claude: {error}"}

    own_errand(process.pid, True)
    try:
        out, err = process.communicate(timeout=PLAN_TIMEOUT)
    except subprocess.TimeoutExpired:
        process.kill()
        process.communicate()
        return {"ok": False, "message": f"Claude did not answer within {int(PLAN_TIMEOUT)}s"}
    except (OSError, subprocess.SubprocessError) as error:
        process.kill()
        return {"ok": False, "message": f"Could not run claude: {error}"}
    finally:
        own_errand(process.pid, False)

    if process.returncode != 0:
        return {"ok": False, "message": git_said(err) or "Claude could not read your usage"}
    if not out.strip():
        return {"ok": False, "message": "Claude answered with nothing"}
    return {"ok": True, **parse_plan(out)}


def read_plan(force: bool = False) -> dict:
    """The account's remaining plan, read at most every few minutes.

    A reading costs five seconds and a process, and the figure moves in
    percentage points over hours, so it is kept and handed back until it is stale.
    Two people opening the dialog at once get the same answer rather than two
    runs: the second is told one is on its way and shown what there is.
    """
    now = time.time()
    with PLAN_LOCK:
        held = dict(PLAN_HELD)
        fresh = held.get("ok") and now - held.get("at", 0) < PLAN_FRESH
        if fresh and not force:
            return {**held, "reading": False}
        if config.PLAN_RUNNING:
            # Somebody's run is already in flight. Hand back what we have and say
            # so, rather than starting a second `claude` for the same answer.
            return {**held, "reading": True} if held else {"ok": False, "reading": True,
                                                          "message": "Reading your usage…"}
        config.PLAN_RUNNING = True

    try:
        answer = run_plan()
    finally:
        with PLAN_LOCK:
            config.PLAN_RUNNING = False

    answer["at"] = time.time()
    if answer.get("ok"):
        with PLAN_LOCK:
            PLAN_HELD.clear()
            PLAN_HELD.update(answer)
        return {**answer, "reading": False}
    # A failed read does not throw away a good one: the last figures with the
    # reason the refresh failed is more use than the reason alone.
    if held.get("ok"):
        return {**held, "reading": False, "message": answer.get("message", "")}
    return {**answer, "reading": False}
